actors/builtin/miner/miner_actor.go

```diff
@@ -1783,19 +1803,53 @@ func validateFRDeclaration(deadlines *Deadlines, deadline *DeadlineInfo, declare
 	return nil
 }
 
-// Computes a fee for a collection of sectors and unlocks it from unvested funds (for burning).
-// The fee computation is a parameter.
-func unlockPenalty(st *State, store adt.Store, currEpoch abi.ChainEpoch, sectors []*SectorOnChainInfo,
-	feeCalc func(info *SectorOnChainInfo) abi.TokenAmount) (abi.TokenAmount, error) {
-	fee := big.Zero()
+// qaPowerForSectors sums the quality adjusted power of all sectors
+func qaPowerForSectors(sectors []*SectorOnChainInfo) (abi.StoragePower, error) {
+	power := big.Zero()
 	for _, s := range sectors {
-		fee = big.Add(fee, feeCalc(s))
+		sectorSize, err := s.Info.SealProof.SectorSize()
+		if err != nil {
+			return abi.NewTokenAmount(0), xerrors.Errorf("could not get sector size for sector: %w", err)
+		}
+		power = big.Add(power, QAPowerForSector(sectorSize, s))
+	}
+	return power, nil
+}
+
+func unlockDeclaredFaultPenalty(st *State, store adt.Store, currEpoch abi.ChainEpoch, epochTargetReward abi.TokenAmount, networkQAPower abi.StoragePower, sectors []*SectorOnChainInfo) (abi.TokenAmount, error) {
+	totalQAPower, err := qaPowerForSectors(sectors)
+	if err != nil {
+		return abi.NewTokenAmount(0), err
+	}
+	fee := PledgePenaltyForSectorDeclaredFault(epochTargetReward, networkQAPower, totalQAPower)
+	return st.UnlockUnvestedFunds(store, currEpoch, fee)
+}
+
+func unlockUndeclaredFaultPenalty(st *State, store adt.Store, currEpoch abi.ChainEpoch, epochTargetReward abi.TokenAmount, networkQAPower abi.StoragePower, sectors []*SectorOnChainInfo) (abi.TokenAmount, error) {
+	totalQAPower, err := qaPowerForSectors(sectors)
+	if err != nil {
+		return abi.NewTokenAmount(0), err
 	}
+	fee := pledgePenaltyForSectorUndeclaredFault(epochTargetReward, networkQAPower, totalQAPower)
 	return st.UnlockUnvestedFunds(store, currEpoch, fee)
 }
 
+func unlockTerminationPenalty(st *State, store adt.Store, curEpoch abi.ChainEpoch, epochTargetReward abi.TokenAmount, networkQAPower abi.StoragePower, sectors []*SectorOnChainInfo) (abi.TokenAmount, error) {
+	totalFee := big.Zero()
+	for _, s := range sectors {
+		sectorSize, err := s.Info.SealProof.SectorSize()
+		if err != nil {
+			return abi.NewTokenAmount(0), xerrors.Errorf("could not get sector size for sector: %w", err)
+		}
+		sectorPower := QAPowerForSector(sectorSize, s)
+		fee := pledgePenaltyForSectorTermination(s, epochTargetReward, networkQAPower, sectorPower)
```

Comment:
This one is going to end up being more complicated, including reference to the sector's stored initial pledge.

Reply:
Yes we'll be sure to add that parameter in the next PR